openwisp-config/files/sbin/openwisp-update-config.lua

```diff
@@ -148,7 +148,7 @@ if lfs.attributes(remote_config_dir, 'mode') == 'directory' then
         -- ensure we are acting on a file
         if lfs.attributes(remote_path, 'mode') == 'file' then
             -- if there's no backup of the file yet, create one
-            if not utils.file_exists(stored_path) then
+            if (not utils.file_exists(stored_path) and not utils.file_exists(remote_path)) then
```

Comment:
I don't understand this change. `remote_path` is the config file coming from the controller. Why should we create a backup of the file on the device if there's no corresponding file coming from the controller?

Reply:
I think this change is too naive to solve the issue.

The remote UCI file may exist both remotely and locally but the two may be totally different and this simple change won't be able to deal with it.

I think we have to compare what comes from the remote conf and what is present locally and do some kind of subtraction:

- take what is present locally
- check if any local conf is also present in remote conf, in that case either avoid to add it or remove it
- store the result

@okraits BTW this solution would not allow us to remove a specific local configuration easily: once a local conf is backed up, it would be restored once the corresponding conf is removed from OpenWISP, the only way to clear it up would be with manual intervention on the device, either via SSH or via send command feature of the new version of OpenWISP Conntroller. I am not sure wheter we should invest further in this functionality right now since it sounds good enough even though not always comfortable to use in 100% of cases. I think over the long run we will have to invest in storing the full conf on OpenWISP which may make managing the conf easier.

Reply:
As I said in https://github.com/openwisp/openwisp-config/issues/153#issuecomment-979845764, we cannot solve this completely:

* We have to look at every individual section in the local configuration
* If it is present in the remote configuration, do not back it up - but **we can't restore its previous state if it was modified by the controller**
* If it isn't preset in the remote configuration, add it to the backup

Yes, we are not able to remove a local configuration unless it was added by the controller. As we recently discussed in the chat, right?

Yes, in the long run we have to migrate from the merge principle to a push-pull principle - that the complete configuration is synchronized between controller and client, as in https://github.com/openwisp/openwisp-config/pull/107.

We are also interested in that and I would help with getting this closer to reality.

Reply:
Correct me if I am wrong, but:

we are looping over the files in `remote_config_dir`, so `not utils.file_exists(remote_path)` will never be true, right?

So I would remove this change completely.